src/rgw/rgw_rest_s3.cc

```diff
@@ -2541,20 +2541,30 @@ void RGWCompleteMultipart_ObjStore_S3::send_response()
   if (op_ret == 0) { 
     dump_start(s);
     s->formatter->open_object_section_in_ns("CompleteMultipartUploadResult", XMLNS_AWS_S3);
-    if (!s->bucket_tenant.empty()) {
-      if (s->info.domain.length()) {
-        s->formatter->dump_format("Location", "%s.%s.%s",
-          s->bucket_name.c_str(),
-          s->bucket_tenant.c_str(),
-          s->info.domain.c_str());
+    std::string domain = (!s->info.domain.empty()) ? s->info.domain :
+      [&]() -> std::string {
+      std::string domain = "http://";
+      if (this->s->info.env->exists("SERVER_NAME")) {
+	domain.append(this->s->info.env->get("SERVER_NAME", "<SERVER_NAME>"));
+      } else {
+	domain.append(this->s->info.env->get("HTTP_HOST", "<HTTP_HOST>"));
       }
+      return domain;
+    }();
```

Comment:
What's the reason for wrapping the code with the immediately executed lambda?

Reply:
just permits lining it into the ternary assignment expr.  most likely this should be consolidated into a routine, but I don't know where all it should be used yet

Reply:
Yeah, having a routine for that sounds nice. We might also want to unify with `dump_uri_from_state` of `rgw_rest.cc`.

Reply:
does my routine (based on the code formerly in this location but adjusted for tenant) do the right thing?  dump_uri_from_state() is different, ignores tenants, deals with domain differently;  need to get it right for all cases...

Reply:
`<Location>` crafted in the new way is definitely more consistent with examples from the AWS doc. The last thing I see is the HTTPS case.

Actually `dump_uri_from_state` looks like a dead code. I will make a patch to drop it.

Reply:
@rzarzynski that's a big help, I'll update this one with the cleanups